docs/codeql/codeql-workshops/cpp/introduction/session-1/codeql-workshop-cpp-bad-overflow-check.md

```diff
@@ -0,0 +1,229 @@
+
+Where might this go wrong?
+
+<details>
+<summary>Answer</summary>
+CPUs will generally prefer to perform arithmetic operations on 32-bit or larger integers, as architectures are optimized to perform these efficiently. The compiler therefore performs "integer promotion" for arguments to arithmetic operations that are smaller than 32-bits.
+
+Unfortunately, this means that any arithmetic operations that occur on types smaller than 32 bits will not overflow as expected, because they have been calculated using a larger type.
+
+The arguments of the following arithmetic operators undergo implicit conversions:
+ * binary arithmetic * / % + -
+ * relational operators < > <= >= == !=
+ * binary bitwise operators & ^ |
+ * the conditional operator ?:
+
+See <a href="https://en.cppreference.com/w/c/language/conversion">https://en.cppreference.com/w/c/language/conversion</a> for more details.
+</details>
+
+Consider this example:
+```c
+uint16_t v = 65535;
+uint16_t b = 1;
+uint16_t result;
+if (v + b < v) {
+    handle_error("overflow");
+} else {
+    result = v + b;
+}
+```
+What is the value of result?
+<details>
+<summary>Answer</summary>
+Here's the example again, with the conversions made explicit:
+
+```c
+uint16_t v = 65535;
+uint16_t b = 1;
+uint16_t result;
+if ((int)v + (int)b < (int)v) {
+    handle_error("overflow");
+} else {
+    result = (uint16_t)((int)v + (int)b);
+}
+```
+
+In this example the second branch is executed, even though there is a 16-bit overflow, and result is set to zero.
+
+Explanation:
+
+ * The two integer arguments to the addition, `v` and `b`, are promoted to 32-bit integers.
+ * The comparison (`<`) is also an arithmetic operation, therefore it will also be completed on 32-bit integers.
+ * This means that `v + b < v` will never be true, because v and b can hold at most 2<sup>16</sup>.
+ * Therefore, the second branch is executed, but the result of the addition is stored into the result variable. Overflow will still occur as result is a 16-bit integer.
+
+In simple terms, if you have an operation `v + b < v`, and `v` and `b` both have a type smaller than 32-bits, then the overflow check will never succeed.
+</details>
+
+We will begin by using CodeQL to identify relational comparison operations (`<`, `>`, `<=`, `>=`) that look like overflow checks. We will then refine this to identify only those checks which are likely wrong due to integer promotion.
+
+If you follow the workshop all the way to the end, then you will find a real security in an old version of ChakraCore.
+
+## Setup instructions
+
+### Writing queries on your local machine
+
+To run CodeQL queries on ChakraCore offline, follow these steps:
+
+1. Install the Visual Studio Code IDE.
+1. Download and install the [CodeQL extension for Visual Studio Code](https://help.semmle.com/codeql/codeql-for-vscode.html). Full setup instructions are [here](https://help.semmle.com/codeql/codeql-for-vscode/procedures/setting-up.html).
```

Comment:
You'll need to change help.semmle.com references to the corresponding codeql.github.com links; apologies that these do not automatically redirect.

Reply:
done